File: netDxf/Entities/HatchSplineData.cs
// Copyright (c) netDxf contributors. Licensed under the MIT License.
using System;

namespace netDxf.Entities
{
    internal static class HatchSplineData
    {
        internal static void Validate(HatchBoundaryPath.Spline spline)
        {
            if (spline.Degree < 1)
                throw new ArgumentException("HATCH spline degree must be positive.");
            if (spline.Knots == null || spline.ControlPoints == null)
                throw new ArgumentException("HATCH spline knots and control points must be supplied.");
            // Periodic wire representations differ from the compressed ObjectARX
            // API representation. Do not impose an unqualified periodic relation.
            if (!spline.IsPeriodic)
            {
                if (spline.ControlPoints.Length < (int)spline.Degree + 1)
                    throw new ArgumentException("A nonperiodic HATCH spline requires at least degree plus one control points.");
                if ((long)spline.Knots.Length != (long)spline.ControlPoints.Length + spline.Degree + 1)
                    throw new ArgumentException("A nonperiodic HATCH spline knot count must equal its control count plus degree plus one.");
            }
            for (int i = 0; i < spline.Knots.Length; i++)
            {
                if (!Finite(spline.Knots[i]))
                    throw new ArgumentException("HATCH spline knots must be finite.");
                if (i > 0 && spline.Knots[i] < spline.Knots[i - 1])
                    throw new ArgumentException("HATCH spline knots must be nondecreasing.");
            }
            foreach (Vector3 point in spline.ControlPoints)
                if (!Finite(point.X) || !Finite(point.Y) || !Finite(point.Z))
                    throw new ArgumentException("HATCH spline control coordinates and stored weights must be finite.");
        }

        private static bool Finite(double value)
        { return !double.IsNaN(value) && !double.IsInfinity(value); }
    }
}

File: javascript/netDxf/Entities/HatchSplineData.js
// Copyright (c) netDxf contributors. MIT License; see package LICENSE.
import { ArgumentException } from '../../runtime/Errors.js';
export class HatchSplineData {
  static Validate(spline) {
    if (spline.Degree < 1) throw new ArgumentException('HATCH spline degree must be positive.');
    if (spline.Knots == null || spline.ControlPoints == null) throw new ArgumentException('HATCH spline knots and control points must be supplied.');
    if (!spline.IsPeriodic) {
      if (spline.ControlPoints.length < spline.Degree + 1) throw new ArgumentException('Insufficient HATCH control points.');
      if (spline.Knots.length !== spline.ControlPoints.length + spline.Degree + 1) throw new ArgumentException('Invalid HATCH knot count.');
    }
    for (let i=0;i<spline.Knots.length;i++) {
      if (!Number.isFinite(spline.Knots[i])) throw new ArgumentException('HATCH knots must be finite.');
      if (i>0 && spline.Knots[i]<spline.Knots[i-1]) throw new ArgumentException('HATCH knots must be nondecreasing.');
    }
    for (const p of spline.ControlPoints) if (![p.X,p.Y,p.Z].every(Number.isFinite)) throw new ArgumentException('HATCH control coordinates and stored weights must be finite.');
  }
}
